modules/client/src/connect.ts

```diff
@@ -0,0 +1,294 @@
+import "core-js/stable";
+import { Contract, providers, utils } from "ethers";
+import { AddressZero } from "ethers/constants";
+import { fromExtendedKey, fromMnemonic } from "ethers/utils/hdnode";
+import tokenAbi from "human-standard-token-abi";
+import "regenerator-runtime/runtime";
+
+import { ChannelProvider, createCFChannelProvider } from "./channelProvider";
+import { ConnextClient } from "./connext";
+import { Logger, stringify } from "./lib";
+import { NodeApiClient } from "./node";
+import {
+  CFCoreTypes,
+  ClientOptions,
+  ConnextClientStorePrefix,
+  CreateChannelMessage,
+  GetConfigResponse,
+  IConnextClient,
+  Store,
+} from "./types";
+
+type KeyGen = (index: string) => Promise<string>;
+
+interface EthProviderSetup {
+  ethProvider: providers.JsonRpcProvider;
+  network: utils.Network;
+}
+
+interface ServiceSetup {
+  messaging: IMessagingService;
+  node: NodeApiClient;
+  config: GetConfigResponse;
+}
+
+interface ChannelProviderOptions {
+  keyGen?: KeyGen;
+  mnemonic?: string;
+  nodeUrl?: string;
+  store?: Store;
+  xpub?: string;
+}
+
+interface ChannelProviderSetup {
+  node: NodeApiClient;
+  messaging: IMessagingService;
+  channelProvider: ChannelProvider;
+  config: GetConfigResponse;
+  keyGen: KeyGen;
+  store: Store;
+}
+
+const setupEthProvider = async (ethProviderUrl: string): Promise<EthProviderSetup> => {
+  const ethProvider = new providers.JsonRpcProvider(ethProviderUrl);
+  const network = await ethProvider.getNetwork();
+
+  // special case for ganache
+  if (network.chainId === 4447) {
+    network.name = "ganache";
+    // Enforce using provided signer, not via RPC
+    ethProvider.getSigner = (addressOrIndex?: string | number): any => {
+      throw { code: "UNSUPPORTED_OPERATION" };
+    };
+  }
+
+  return { ethProvider, network };
+};
+
+const createMessagingService = async (
+  nodeUrl: string,
+  logLevel: number,
+): Promise<IMessagingService> => {
+  const messagingFactory = new MessagingServiceFactory({
+    logLevel,
+    messagingUrl: nodeUrl,
+  });
+  const messaging = messagingFactory.createService("messaging");
+  await messaging.connect();
+  return messaging;
+};
+
+const setupServices = async (
+  nodeUrl: string,
+  log: Logger,
+  logLevel: number,
+): Promise<ServiceSetup> => {
+  // create a messaging service client
+  log.debug(`Creating messaging service client (logLevel: ${logLevel})`);
+  const messaging = await createMessagingService(nodeUrl, logLevel);
+
+  // create a new node api instance
+  const node = new NodeApiClient({ logLevel, messaging });
+  const config = await node.config();
+  log.debug(`Node provided config: ${stringify(config)}`);
+  return { messaging, node, config };
+};
+
+const setupChannelProvider = async (
+  ethProvider: providers.JsonRpcProvider,
+  log: Logger,
+  logLevel: number,
+  providedChannelProvider?: ChannelProvider,
+  channelProviderOptions?: ChannelProviderOptions,
+): Promise<ChannelProviderSetup> => {
```

Comment:
My rule of thumb: don't save anything to a variable that we only use once; just use it in-place.

This connect function was kind of a monolith so there's a valid argument that breaking it out into bite-size pieces increases readability.

But counterargument would be that I can't read the connect function straight-through anymore. To understand what it's doing, I have to constantly track down the extracted functions (+ the extra types that were added) which is harder for me to read than one big function where everything I need is in once place.

(Obviously, if there's logic we need to re-use, then not repeating ourselves trumps any readability concerns)

I don't have super strong feelings here though, feel free to leave this as-is if you like this style better.

Reply:
This will most likely to change and it was also mostly driven to control the scope of some variables which were shared across the whole connect method. Also the idea is to port some of this logic into the channelProvider inside the enable method.